Sidebar: WIP Refactor of sidebar CSS

shaunandrews opened the pull request:
WIP

Ok, so this is gonna be a big one, but I thinks a long time coming. The Calypso sidebar has been redesigned recently, but there's some legacy code still lingering around complicating things. The recent redesign overrode existing styles, rather than replaced them.

The goal is to introduce very few visual changes, but reduce and simplify the code where possible. So far, the visual changes are:

- Updating the expandable arrow icon (chevron) with a material icon.
- Move the expandable arrow icon (chevron) for Reader's Lists and Tags to the right.
- Make all menu item's font-weight normal, but expandable items bold.
- "Unstick" the WP Admin and Add New Site menu items from the bottom of the sidebar.

Here's a Before/After for My Site(s) with multiple sites and a single site:

<img width="1524" alt="image" src="https://user-images.githubusercontent.com/191598/60524738-1b290000-9cbb-11e9-8767-70b4b32cda08.png">

shaunandrews committed:
Added a new icon and using it for the Reader sidebar conversations menu item.

shaunandrews committed:
Taking a stab at cleaning up the sidebar CSS, with a focus around the expandable-heading pattern.

shaunandrews committed:
Having this an a UL complicates things, and makes it messy when we have child elements not within an li.

shaunandrews committed:
This link didn't have the same class as other links. Adding the class lets us avoid having special CSS for this.

shaunandrews committed:
We no longer show these icons. We were hiding them with CSS, but this is better.

shaunandrews committed:
No more icons for these items.

shaunandrews committed:
Refactoring in progress. Its a mess, I know. Have faith.

shaunandrews committed:
I think we've been hiding this icon with CSS. Lets just remove it from the markup.

shaunandrews committed:
Adding a classname to the plan and wpadmin icons to make them easier to style.

shaunandrews committed:
Updating a few classnames to make things a bit easier to style.

shaunandrews committed:
Replacing the ul with a div, updating a classname, and moving the add form.

Not sure about this addform in general.

shaunandrews committed:
Adding a classname to make icons easier to style.

shaunandrews committed:
Making progress. Starting to look like the production sidebar. Lots of edgecases left.

shaunandrews committed:
Some SVG hovers and adding the secondary text styles.

shaunandrews committed:
Positioning external icons.

shaunandrews committed:
Switching the add new site from a button to a sidebarItem.

shaunandrews committed:
Add some spacing and a border to the "wp admin" and "add new" menu items.

shaunandrews committed:
Removing some unused CSS

shaunandrews committed:
More tweaks I forgot about.

a8ci18n commented on client/my-sites/sidebar/sidebar.jsx:
```diff
@@ -584,14 +575,12 @@ export class MySitesSidebar extends Component {
-				<Gridicon icon="add-outline" /> { this.props.translate( 'Add New Site' ) }
-			</Button>
+			<SidebarItem
+				label={ this.props.translate( 'Add new site' ) }
```
Hi! I've found a possible matching string that has already been translated **22** times:
`translate( '+ Add New Site' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/sidebar-CSS-refactor)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/reader/sidebar/index.jsx:
```diff
@@ -228,10 +219,12 @@ export class ReaderSidebar extends React.Component {
-									<span className="menu-link-text">{ this.props.translate( 'My Likes' ) }</span>
-								</a>
+								<SidebarItem
+									label={ this.props.translate( 'My likes' ) }
```
:disappointed: 61 existing translations will be lost with this change.
If the string must change, the following option has already been translated **30** times:
`translate( 'Go to My Likes' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

fditrapani committed:
Update sidebbar heading

fditrapani committed:
Fix nested lists and selected states

a8ci18n commented on client/reader/sidebar/index.jsx:
```diff
@@ -131,108 +130,86 @@ export class ReaderSidebar extends React.Component {
 
-							<li
+							<SidebarItem
+								label={ this.props.translate( 'My likes' ) }
```
Hi! I've found a possible matching string that has already been translated **61** times:
`translate( 'My Likes' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/sidebar-CSS-refactor)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

fditrapani committed:
Fix Automattic icons

fditrapani committed:
Add icon for tags

fditrapani committed:
Fix reader tags

jsnajdr commented:
Hi @shaunandrews :wave: How finished or unfinished is this branch? If we rebase it onto the latest `master` and resolve the Git conflicts, is it ready to be reviewed and merged? Or are there any major sub-tasks to be done?

We're migrating our CSS files to webpack and finishing this PR would be a major step towards finishing that migration project.

jsnajdr committed:
Unify colors of visited and non-visited sidebar links

jsnajdr committed:
Remove duplicate icon for Stats

jsnajdr committed:
Make the font size of sidebar header the same as sidebar item (was smaller)

jsnajdr committed:
Improve the useWPAdminFlows logic

jsnajdr commented:
Just rebased this onto latest `master`, resolved conflicts and pushed a few little fixes.

Some issues to resolve:
- this PR maybe breaks or overwrites a sidebar item focus ring 1px glitch that @adamziel fixed in #35138. Adam, can you verify the status of that glitch in this branch?
- another PR that probably breaks is wp-admin link color fix from #34986. @Aurorum can you verify that the new wp-admin link is OK?

The "Add tag" UI in Reader needs some fixing, too:
<img width="277" alt="Screenshot 2019-08-23 at 15 25 52" src="https://user-images.githubusercontent.com/664258/63598565-1f3d1400-c5c0-11e9-8921-54c1b6507f2e.png">
- there used to be a `Count` of tags as part of the sidebar item, and it's gone now
- the "Add" input should be reset back to empty content (or disappear) after submitted

Cc @diegohaz: this is related to the Sidebar CSS migration.

Aurorum commented:
> can you verify that the new wp-admin link is OK?

(I'm assuming you refer to the "Add New Site" link which that PR intended to fix)

Afraid it isn't okay :(

Hover state for "Add New Site" on a single site in Nightfall (this isn't exclusive to the "Add New Site" link though, it affects all links in the sidebar):
<img width="268" alt="Screenshot 2019-08-23 at 15 37 26" src="https://user-images.githubusercontent.com/43215253/63600727-03d00a00-c5bc-11e9-9c8a-fd9642715b0b.png">

And on the Ocean scheme:

<img width="274" alt="Screenshot 2019-08-23 at 15 39 39" src="https://user-images.githubusercontent.com/43215253/63600837-3e39a700-c5bc-11e9-9eab-c9ea35107f9e.png">

Also affects the other schemes such as Sunset. For an account with multiple sites, this PR seems fine for the "Add New Site" button on all schemes.

But for the wp-admin link, the icon disappears on hover (scheme: Sunset)

<img width="270" alt="Screenshot 2019-08-23 at 15 43 46" src="https://user-images.githubusercontent.com/43215253/63601181-db94db00-c5bc-11e9-916e-d017fe3b8901.png">

The toggles also aren't changing colour on different schemes, they're just remaining black

shaunandrews commented:
> there used to be a Count of tags as part of the sidebar item, and it's gone now

I intentionally removed it, as I found it visually complicated and my gut tells me its rarely useful. We can bring it back if there's an objection.

jsnajdr committed:
Reader Tags: reset add form after submitting

jsnajdr committed:
Reader Tags: fix CSS class names, remove unused styles

jsnajdr committed:
ExpandableHeader: use JSX a bit better

jsnajdr committed:
Remove double icon (grid and material) from Store item

jsnajdr committed:
Fix onHover text color of sidebar item

jsnajdr committed:
Add theme-specific color to sidebar expand icon

jsnajdr committed:
StatsSparkline: useTranslate hook, use Fragment instead of wrapper span

jsnajdr committed:
Woo: fix style of count badge in sidebar

jsnajdr committed:
Sidebar: use JSX a bit better

jsnajdr committed:
SidebarItem: render children inside the anchor, expand label with flex

Makes it easier to insert right-aligned special markup to sidebar items:
plan name, stats sparkline, Woo order counts... And to avoid absolute
positioning in favor of flex.

jsnajdr committed:
Fix and improve positioning of StatsSparkline: flex instead of absolute

jsnajdr committed:
FormTextInputWithAction: don't pass defaultValue to controlled input component

The `defaultValue` is used to initialize state. Passing the prop to a fully controlled
(i.e., with `value` prop) subcomponent triggers a React warning about controlled vs
uncontrolled component confusion.

jsnajdr commented:
> I intentionally removed it, as I found it visually complicated and my gut tells me its rarely useful. We can bring it back if there's an objection.

@shaunandrews No problem, just wanted to confirm that it was the intent 🙂 

I pushed a few more updates with several fixes:

Fixed hover styles in Sunset and Ocean themes, as reported by @Aurorum 

Fixed color of expansion icon: use theme text color instead of black. Buggy state:
<img width="277" alt="Screenshot 2019-08-27 at 10 38 19" src="https://user-images.githubusercontent.com/664258/63765014-7268dc80-c8c8-11e9-97e8-d595c886fa64.png">

Fixed styling of the stats sparkline. Due to absolute positioning and broken CSS styles, it was rendered completely outside the Stats item:
<img width="391" alt="Screenshot 2019-08-27 at 11 00 50" src="https://user-images.githubusercontent.com/664258/63765095-99bfa980-c8c8-11e9-83b2-73d8a53f284d.png">
I fixed that by abandoning `position: absolute` and using Flex instead.

Fixed the sidebar in Woo Store, especially the count badges:
<img width="276" alt="Screenshot 2019-08-27 at 11 54 58" src="https://user-images.githubusercontent.com/664258/63765167-c2e03a00-c8c8-11e9-9584-74d04be5ef56.png">

As far as I'm concerned, this is now ready to merge 🙂

jsnajdr committed:
ReaderSidebarLists: fix lint error

Aurorum commented:
Hi @jsnajdr, I spotted two more bugs which this PR would introduce. Firstly, on different colour schemes, the plan name isn't using the right colour variable. On the Ocean scheme: 

<img width="279" alt="Screenshot 2019-08-27 at 13 32 43" src="https://user-images.githubusercontent.com/43215253/63771487-2a00ed00-c8cf-11e9-8fcb-378c2e153418.png">

Secondly (also on Ocean), when I've selected a tab, I can hover over the icon and it'll change colour. This doesn't occur currently. 

<img width="260" alt="Screenshot 2019-08-27 at 13 33 57" src="https://user-images.githubusercontent.com/43215253/63771572-561c6e00-c8cf-11e9-8528-2884b004d084.png">

jsnajdr committed:
Fix onHover and selected styles of icons and secondary text

jsnajdr commented:
Thanks for reporting @Aurorum! Both issues fixed in c7abc02.

Aurorum commented:
In the Contrast scheme, hovering over the "Plan" tab or a toggle heading renders the plan name/heading title unreadable (doesn't currently happen):

<img width="281" alt="Screenshot 2019-08-27 at 18 57 12" src="https://user-images.githubusercontent.com/43215253/63795773-90e8cb00-c8fc-11e9-8c63-4f0926ebbe9a.png">
<img width="275" alt="Screenshot 2019-08-27 at 18 56 33" src="https://user-images.githubusercontent.com/43215253/63795772-90503480-c8fc-11e9-9325-31d6111a6b3d.png">

jsnajdr committed:
Fix and reshuffle the hover and selected styles for icons and text

jsnajdr commented:
Thanks again @Aurorum for thorough testing through all the themes! Another set of CSS fixes for hover and selected states is in 1998940.

Aurorum commented on client/layout/sidebar/style.scss:
```diff
@@ -1,731 +1,206 @@
-		float: left;
-		margin-right: 6px;
+	// Some links are marked as "external" and show
+	// and icon to visually indicate they will open
```
```suggestion
	// an icon to visually indicate they will open
```

Aurorum commented on client/layout/sidebar/style.scss:
```diff
@@ -1,731 +1,206 @@
+		border-bottom: 1px solid var( --sidebar-border-color );
 
-		@include breakpoint( '<660px' ) {
-			border-top: none;
```
I'm not so sure about getting rid of this, it seems to currently prevent there being a double border on mobile.

<img width="344" alt="Screenshot 2019-08-28 at 08 52 59" src="https://user-images.githubusercontent.com/43215253/63836364-54a38200-c971-11e9-9bb0-1003788f03f2.png">

But with this branch, there is a double border, and I'm assuming that's not intentional. 

<img width="349" alt="Screenshot 2019-08-28 at 08 52 53" src="https://user-images.githubusercontent.com/43215253/63836381-608f4400-c971-11e9-80c4-e79c3c03266d.png">

jsnajdr replied:
I decided to remove these mobile-only borders. It's hard to make them right in all scenarios, and they are not terribly useful. Other popular mobile apps (looked at Slack and Instagram) don't have borders on menu items and still are great.

Aurorum replied:
Agreed, it doesn't seem necessary and it still looks good.

Aurorum commented on client/reader/sidebar/index.jsx:
```diff
@@ -128,111 +127,105 @@ export class ReaderSidebar extends React.Component {
 
-							<li
+							<SidebarItem
+								label={ this.props.translate( 'My likes' ) }
```
Minor, but it's currently "My Likes" and that feels like the more consistent option (for example, it's currently "Followed **S**ites".

```suggestion
								label={ this.props.translate( 'My Likes' ) }
```

<img width="205" alt="Screenshot 2019-08-28 at 08 58 20" src="https://user-images.githubusercontent.com/43215253/63836726-00e56880-c972-11e9-92b9-88120154eee5.png">

Aurorum reviewed (commented):
This PR is testing well, just a few comments though! :)

jsnajdr committed:
Fix typo in CSS comment

jsnajdr committed:
Reader: capitalize My likes to My Likes, improve JSX

jsnajdr committed:
Remove mobile-only borders around sidebar items

jsnajdr committed:
Classic Editor Confirmation Sidebar: remove style dependency on Sidebar, fix styles

Removes the dependency of the Confirmation Sidebar styles on the main Sidebar and implements
its own independent styling (not very hard) and fixes a few issues.

jsnajdr commented:
@Aurorum I implemented all your suggestions:
- fixed a typo in CSS comment
- removed mobile-only borders
- capitalized "My Likes" in Reader sidebar

I also discovered that the `Sidebar` styles were reused (with SASS `import` and `extend`) for the `EditorConfirmationSidebar` in Classic Editor UI:

<img width="277" alt="Screenshot 2019-08-28 at 12 19 08" src="https://user-images.githubusercontent.com/664258/63847986-38aed900-c98f-11e9-91cd-8a293a88c18d.png">

I had to do a few fixes in 9a4f19f to put it back to order.

jsnajdr reviewed (approved):
OK, time to merge! :tada:

Aurorum reviewed (approved):
I've extensively tested:
- the sidebar at all the different breakpoints 
- the sidebar in Me, My Sites, DevDocs, Shop (though I can't access much there), the classic editor, and the Reader
- the sidebar in all the different colour schemes
- the sidebar for simple sites and Jetpack sites (where there's an external icon)
- the sidebar for both users with one site and multiple sites

Besides the bugs above which have now all been fixed, everything works perfectly as expected!
